core/commands/ls.go

```diff
@@ -21,24 +21,34 @@ import (
 	"gx/ipfs/Qmde5VP1qUkyQXKCfmEUA7bP64V2HAptbJ7phuPp7jXWwg/go-ipfs-cmdkit"
 )
 
+// LsLink contains printable data for a single ipld link in ls output
 type LsLink struct {
 	Name, Hash string
 	Size       uint64
 	Type       unixfspb.Data_DataType
 }
 
+// LsObject is an element of LsOutput
+// It can represent a whole directory, a directory header, one or more links,
+// Or a the end of a directory
 type LsObject struct {
-	Hash  string
-	Links []LsLink
+	Hash      string
+	Links     []LsLink
+	HasHeader bool
+	HasLinks  bool
+	HasFooter bool
```

Comment:
I would like to avoid the need for the extra flags.  We should instead just send data as we get it and let the client figure out the rest.

If a flag at this level is really needed I wold like to use something like `Incomplete` which will signal to the client that not all the links for the given hash has been delivered.